### DISCUSSION TITLE
Audio Switching

### DISCUSSION THREAD

**Discussion by Ayaki683** [ClassicOldSong/Apollo#111] [Q&A] [answered]
Is there any usable projects to switch audio devices automaticly for only one software which is streaming,while others are not be infused?

**Comment by ClassicOldSong** [comment]
There's no to my knowledge, as the device in use is determined by the app that is playing audio itself.

**Reply by Ayaki683** [reply]
But I've found a solution that using a third party software like [Audio Router](https://github.com/audiorouterdev/audio-router),it can change the audio playing through another speaker which is not determined by the app itself,but determined to which you choices.Unfortunately, it not provides command, but to use a GUI. And Sunshine can't catch the audio with my another speaker choice.,it can only catch with the same choice in Windows.

**Reply by ClassicOldSong** [reply]
It seems it's not working properly in Windows 11 and somehow requires a signature.

In Apollo you can uncheck `Keep virtual sink as default` and `Auto capture current sink`, and specify the audio output you want to capture in `Virtual Sink`.

**Reply by Ayaki683** [reply]
This is a solution.

> In Apollo you can uncheck `Keep virtual sink as default` and `Auto capture current sink`, and specify the audio output you want to capture in `Virtual Sink`.

But I need to pre-set the app's audio in `Setting>Volume>Volume Mixer`,and set the speaker back manually when launching the app(Because it set my virtual speaker automately). Is there any command to make these steps automately,or just catch my virtual speaker without set it as default? Maybe [Audio Switcher](https://github.com/xenolightning/AudioSwitcher/tree/master) can help.

**Reply by ClassicOldSong** [reply]
IDK if there's commands available for this. Audio routing/mixing have various conflicting cases where some peopoe want to capture specific output, some people want to capture specific application, and some peopoe want to capture them all, while some other people want to auto capture default output whether what it is right now. It's very very hard to get right with so many conflicting requirements with one single solution. I tried to make the original mess a bit clear with the two additional options I added in Apollo, but more specific needs might need to be solved by your own creative thinking.

**Reply by Ayaki683** [reply, marked as answer]
So,this is my solution.
1.Run `Install-Module -Name AudioDeviceCmdlets -Scope CurrentUser` in powershell.Pre-set the app's audio to `Steam Streming Speakers` in `Setting>Volume>Volume Mixer`. Uncheck `Keep virtual sink as default` and `Auto capture current sink` in Apollo,uncheck `Play host sound` in Moonlight 
2.Use `Get-AudioDevice -List` to View the audio devices on the system,note the index of your **Default Speaker**. **Not** Steam Streming Speakers.
3.Add `powershell -WindowStyle Hidden Start-Sleep -Milliseconds 10000; Set-AudioDevice -Index 4` (For example) into `Applications>Detached Commands`. The Milliseconds depends on the time your streaming launched,the number after `-Index` is what you noted.
4.Enjoy.

**Reply by Ayaki683** [reply]
If it works.Could you just add it on Wiki?

**Reply by ClassicOldSong** [reply]
> If it works.Could you just add it on Wiki?

Sure, glad you found your solution!

I'll clear it a bit before adding it to the wiki.
